torch/optim/adam.py

```diff
@@ -108,6 +120,11 @@ def _init_group(
                     max_exp_avg_sqs.append(state['max_exp_avg_sq'])
                 if group['differentiable'] and state['step'].requires_grad:
                     raise RuntimeError('`requires_grad` is not supported for `step` in differentiable mode')
+
+                # Foreach without capturable does not support a tensor lr
```

Comment:
why does capturable impact whether or not a tensor LR is supported in foreach? wouldn't it work just not be performant?

Reply:
The capturable path allows lr to be a tensor cuz we have the overloads for those ops. The non capturable path uses python math instead as that is more performant without needing as much device memory. An easy way to "support" tensor LR is just to autoswitch to capturable when we discover that lr is a tensor, but that's not necessarily a decision i want to make in this PR.